File: sdk/python/sdk/bridge.py
"""ASGI <-> gRPC reverse-tunnel bridge for Python extensions.

Core delivers HTTP requests over a bidirectional gRPC stream. This module
drives a single stream using an outgoing :class:`asyncio.Queue` as the request
iterator (so we can write responses back onto the same call), translates each
``HttpRequestChunk`` into an ASGI ``scope``/``receive``/``send`` cycle against
the user's FastAPI app, and pushes the captured response back as a
``HttpResponseChunk``.
"""

import asyncio
from typing import Any

import grpc

from .context import UserContext, _user_context
from .proto import tunnel_pb2, tunnel_pb2_grpc


class Config:
    def __init__(
        self,
        extension_key: str,
        core_grpc_url: str,
        is_dev: bool = True,
        version: str = "1.0.0",
        dev_frontend_url: str = "",
    ):
        self.extension_key = extension_key
        self.core_grpc_url = core_grpc_url
        self.is_dev = is_dev
        self.version = version
        self.dev_frontend_url = dev_frontend_url


def start_sdk(app: Any, config: Config) -> None:
    """Blocking entrypoint: connect, register and serve forever."""
    asyncio.run(_run(app, config))


async def _run(app: Any, config: Config) -> None:
    while True:
        try:
            async with grpc.aio.insecure_channel(config.core_grpc_url) as channel:
                stub = tunnel_pb2_grpc.ExtensionTunnelStub(channel)
                out_queue: "asyncio.Queue[tunnel_pb2.TunnelMessage]" = asyncio.Queue()

                # First outgoing message registers the extension.
                await out_queue.put(
                    tunnel_pb2.TunnelMessage(
                        register_req=tunnel_pb2.RegisterRequest(
                            extension_key=config.extension_key,
                            version=config.version,
                            is_dev=config.is_dev,
                            dev_frontend_url=config.dev_frontend_url,
                        )
                    )
                )

                async def request_iter():
                    while True:
                        msg = await out_queue.get()
                        yield msg

                call = stub.Connect(request_iter())
                async for msg in call:
                    if msg.HasField("register_resp"):
                        if not msg.register_resp.success:
                            print(f"registration rejected: {msg.register_resp.error_message}")
                            break
                        print("registration success")
                    elif msg.HasField("http_req_chunk"):
                        asyncio.create_task(
                            _dispatch_asgi(msg.http_req_chunk, app, out_queue)
                        )
                    elif msg.HasField("pong"):
                        pass
        except Exception as e:  # noqa: BLE001 - reconnect on any failure
            print(f"connection error: {e}. retrying in 2s...")
            await asyncio.sleep(2)


async def _dispatch_asgi(req_chunk: Any, app: Any, out_queue: "asyncio.Queue") -> None:
    headers = dict(req_chunk.headers)
    scope = {
        "type": "http",
        "asgi": {"version": "3.0", "spec_version": "2.1"},
        "http_version": "1.1",
        "method": req_chunk.method,
        "scheme": "http",
        "path": req_chunk.path,
        "raw_path": req_chunk.path.encode("utf-8"),
        "query_string": req_chunk.query.encode("utf-8"),
        "root_path": "",
        "headers": [
            (k.lower().encode("utf-8"), v.encode("utf-8")) for k, v in headers.items()
        ],
        "client": ("127.0.0.1", 0),
        "server": ("core", 80),
    }

    # Inject authenticated user from Core-provided headers.
    user_id = headers.get("X-User-Id")
    if user_id:
        roles = [r for r in headers.get("X-User-Roles", "").split(",") if r]
        perms = [p for p in headers.get("X-User-Permissions", "").split(",") if p]
        _user_context.set(UserContext(user_id=user_id, roles=roles, permissions=perms))

    state = {"status": 200, "headers": {}, "body": bytearray()}
    body_sent = False

    async def receive():
        nonlocal body_sent
        if not body_sent:
            body_sent = True
            return {
                "type": "http.request",
                "body": bytes(req_chunk.body_chunk),
                "more_body": False,
            }
        # No further body; keep the app from hanging if it polls again.
        return {"type": "http.request", "body": b"", "more_body": False}

    async def send(message):
        if message["type"] == "http.response.start":
            state["status"] = message["status"]
            state["headers"] = {
                k.decode("utf-8"): v.decode("utf-8") for k, v in message["headers"]
            }
        elif message["type"] == "http.response.body":
            state["body"].extend(message.get("body", b""))
            if not message.get("more_body", False):
                await out_queue.put(
                    tunnel_pb2.TunnelMessage(
                        http_resp_chunk=tunnel_pb2.HttpResponseChunk(
                            stream_id=req_chunk.stream_id,
                            is_first=True,
                            is_last=True,
                            status_code=state["status"],
                            headers=state["headers"],
                            body_chunk=bytes(state["body"]),
                        )
                    )
                )

    try:
        await app(scope, receive, send)
    except Exception as e:  # noqa: BLE001 - return 500 instead of dropping the stream
        await out_queue.put(
            tunnel_pb2.TunnelMessage(
                http_resp_chunk=tunnel_pb2.HttpResponseChunk(
                    stream_id=req_chunk.stream_id,
                    is_first=True,
                    is_last=True,
                    status_code=500,
                    headers={"Content-Type": "text/plain"},
                    body_chunk=f"internal error: {e}".encode("utf-8"),
                )
            )
        )

File: sdk/go/bridge.go
package sdk

import (
	"bytes"
	"context"
	"io"
	"log"
	"net/http"
	"strings"
	"sync"
	"time"

	"github.com/taills/moduless/manifest"
	pb "github.com/taills/moduless/proto/tunnel"
	"google.golang.org/grpc"
	"google.golang.org/grpc/credentials/insecure"
	"google.golang.org/grpc/metadata"
)

// extKeyMetadataField mirrors the Core interceptor's expected metadata key.
const extKeyMetadataField = "x-extension-key"

// Shared client handles initialised by Start, reusable by DB / Files / Events.
var (
	conn   *grpc.ClientConn
	DB     *DBClient
	Files  *FilesClient
	Events *EventClient
)

// clientStream serializes concurrent Send calls on the gRPC client stream.
type clientStream struct {
	stream pb.ExtensionTunnel_ConnectClient
	mu     sync.Mutex
}

func (c *clientStream) Send(msg *pb.TunnelMessage) error {
	c.mu.Lock()
	defer c.mu.Unlock()
	return c.stream.Send(msg)
}

// Start dials Core, registers the extension, and bridges tunnelled HTTP
// requests into the supplied handler. It blocks and reconnects forever.
func Start(handler http.Handler, cfg Config) {
	var err error
	conn, err = grpc.NewClient(cfg.CoreGrpcURL,
		grpc.WithTransportCredentials(insecure.NewCredentials()),
		grpc.WithUnaryInterceptor(extKeyUnaryInterceptor(cfg.ExtensionKey)),
		grpc.WithStreamInterceptor(extKeyStreamInterceptor(cfg.ExtensionKey)),
	)
	if err != nil {
		log.Fatalf("failed to dial Core gRPC: %v", err)
	}
	defer conn.Close()

	// Initialise auxiliary service clients over the shared connection.
	DB = NewDBClient(conn)
	Files = NewFilesClient(conn)
	Events = NewEventClient(conn)

	client := pb.NewExtensionTunnelClient(conn)
	version := cfg.Version
	if version == "" {
		version = "1.0.0"
	}

	registerReq := &pb.RegisterRequest{
		ExtensionKey:   cfg.ExtensionKey,
		Version:        version,
		IsDev:          cfg.IsDev,
		DevFrontendUrl: cfg.DevFEUrl,
	}
	if cfg.ManifestPath != "" {
		if m, err := manifest.Load(cfg.ManifestPath); err != nil {
			log.Printf("warning: failed to load manifest %s: %v", cfg.ManifestPath, err)
		} else {
			applyManifest(registerReq, m)
		}
	}

	// In production mode the SDK ships the built micro-frontend to Core so the
	// gateway can serve it. The zip is built once and re-streamed on reconnect.
	var frontendZip []byte
	if !cfg.IsDev {
		if cfg.FrontendDir != "" {
			data, sum, err := buildFrontendZip(cfg.FrontendDir)
			if err != nil {
				log.Fatalf("failed to bundle frontend: %v", err)
			}
			frontendZip = data
			registerReq.ZipFileSize = uint64(len(data))
			registerReq.ZipSha256 = sum
		} else {
			// No frontend bundled: register as dev so Core completes immediately
			// instead of waiting for a zip that will never arrive.
			log.Printf("no FrontendDir set with IsDev=false; registering without a micro-frontend")
			registerReq.IsDev = true
		}
	}

	for {
		stream, err := client.Connect(context.Background())
		if err != nil {
			log.Printf("tunnel connect failed: %v, retrying in 2s...", err)
			time.Sleep(2 * time.Second)
			continue
		}

		cs := &clientStream{stream: stream}
		if err := cs.Send(&pb.TunnelMessage{
			Payload: &pb.TunnelMessage_RegisterReq{RegisterReq: registerReq},
		}); err != nil {
			log.Printf("registration send failed: %v", err)
			time.Sleep(2 * time.Second)
			continue
		}

		// Stream the bundled micro-frontend, then signal completion so Core
		// extracts it and replies with the registration result.
		if len(frontendZip) > 0 {
			if err := uploadFrontendZip(cs, frontendZip); err != nil {
				log.Printf("frontend upload failed: %v, reconnecting in 2s...", err)
				time.Sleep(2 * time.Second)
				continue
			}
		}

		if err := handleTunnel(cs, handler); err != nil {
			log.Printf("tunnel connection lost: %v. Reconnecting in 2s...", err)
			time.Sleep(2 * time.Second)
		}
	}
}

// applyManifest copies manifest collection/slot declarations into the
// registration request.
func applyManifest(req *pb.RegisterRequest, m *manifest.Manifest) {
	if m.Weight > 0 {
		req.Weight = int32(m.Weight)
	}
	for _, c := range m.Database.Collections {
		col := &pb.CollectionSchema{Name: c.Name}
		for _, idx := range c.Indexes {
			col.Indexes = append(col.Indexes, &pb.IndexSchema{
				Fields: idx.Fields,
				Unique: idx.Unique,
			})
		}
		req.Collections = append(req.Collections, col)
	}
	for _, s := range m.UISlots {
		req.Slots = append(req.Slots, &pb.SlotSchema{
			SlotName:       s.SlotName,
			ComponentEntry: s.ComponentEntry,
		})
	}
	if m.Menu.Icon != "" {
		req.MenuIcon = m.Menu.Icon
	}
	if m.Menu.Path != "" {
		req.MenuPath = m.Menu.Path
	}
	if m.DisplayName != "" {
		req.DisplayName = m.DisplayName
	}
}

func handleTunnel(cs *clientStream, handler http.Handler) error {
	for {
		msg, err := cs.stream.Recv()
		if err != nil {
			return err
		}

		switch payload := msg.Payload.(type) {
		case *pb.TunnelMessage_RegisterResp:
			if !payload.RegisterResp.Success {
				log.Printf("registration rejected: %s", payload.RegisterResp.ErrorMessage)
				return io.EOF
			}
			log.Printf("registration success")

		case *pb.TunnelMessage_HttpReqChunk:
			go serveBridgedRequest(cs, payload.HttpReqChunk, handler)

		case *pb.TunnelMessage_Pong:
			// Heartbeat acknowledged.
		}
	}
}

func serveBridgedRequest(cs *clientStream, reqChunk *pb.HttpRequestChunk, handler http.Handler) {
	target := reqChunk.Path
	if reqChunk.Query != "" {
		target += "?" + reqChunk.Query
	}
	req, err := http.NewRequest(reqChunk.Method, target, bytes.NewReader(reqChunk.BodyChunk))
	if err != nil {
		log.Printf("failed to reconstruct request: %v", err)
		return
	}
	for k, v := range reqChunk.Headers {
		req.Header.Set(k, v)
	}

	// Parse user headers and inject context.
	if userID := req.Header.Get("X-User-Id"); userID != "" {
		userCtx := &UserContext{
			UserID:      userID,
			Roles:       splitNonEmpty(req.Header.Get("X-User-Roles")),
			Permissions: splitNonEmpty(req.Header.Get("X-User-Permissions")),
		}
		req = req.WithContext(context.WithValue(req.Context(), userContextKey, userCtx))
	}

	w := newMockResponseWriter()
	handler.ServeHTTP(w, req)

	headers := make(map[string]string)
	for k, v := range w.Header() {
		if len(v) > 0 {
			headers[k] = v[0]
		}
	}

	_ = cs.Send(&pb.TunnelMessage{
		Payload: &pb.TunnelMessage_HttpRespChunk{
			HttpRespChunk: &pb.HttpResponseChunk{
				StreamId:   reqChunk.StreamId,
				IsFirst:    true,
				IsLast:     true,
				StatusCode: int32(w.statusCode),
				Headers:    headers,
				BodyChunk:  w.body.Bytes(),
			},
		},
	})
}

func splitNonEmpty(s string) []string {
	if s == "" {
		return nil
	}
	return strings.Split(s, ",")
}

// extKeyUnaryInterceptor attaches the extension identity to outgoing unary RPCs.
func extKeyUnaryInterceptor(extKey string) grpc.UnaryClientInterceptor {
	return func(ctx context.Context, method string, req, reply interface{}, cc *grpc.ClientConn, invoker grpc.UnaryInvoker, opts ...grpc.CallOption) error {
		ctx = metadata.AppendToOutgoingContext(ctx, extKeyMetadataField, extKey)
		return invoker(ctx, method, req, reply, cc, opts...)
	}
}

// extKeyStreamInterceptor attaches the extension identity to outgoing streams.
func extKeyStreamInterceptor(extKey string) grpc.StreamClientInterceptor {
	return func(ctx context.Context, desc *grpc.StreamDesc, cc *grpc.ClientConn, method string, streamer grpc.Streamer, opts ...grpc.CallOption) (grpc.ClientStream, error) {
		ctx = metadata.AppendToOutgoingContext(ctx, extKeyMetadataField, extKey)
		return streamer(ctx, desc, cc, method, opts...)
	}
}
